OSM query improvements (#2) and OsmHelper refactoring (#31);Fixed data creation (#35)

xamanu opened the pull request:
Hello Thorsten,

I fear you are not going to be very happy.... Please excuse me, despite of our general strategy, this time I was not able to file a _small_ pull request! :cry: I'm really sorry, but I saw no real possibilities in doing the query improvements (#2) without restructuring OsmHelper (#31) or vice-versa. Everything was too much interconnected.

I still made it as small as possible, which means it's focused only on those two issues, and I was not touching data structure and other open issues. Most of the work has been done inside the `OsmHelper` class - that was somehow the magic line I didn't want to cross.

The code now comes with comments and documentation and I hope this makes it more understandable. At least to me, it seems to be much more structured now and a huge improvement. See the `OsmHelper` class:

**Public functions**
- `get_routes()`\- Get routes, no matter if they are already cached or not
- `get_stops()` - Get stops, no matter if they are already cached or not

**Private (internal) functions:**
- `_query_routes()` - Query OSM for routes
- `_query_stops()` - Query OSM for stops
- `_build_route_master()` - Create RouteMaster object
- `_build_route_variant()` - Create Route(Variant) object
- `_build_stop()` - Create Stop object
- `_cache_write()` - One single function that writes object to the file cache
- `_cache_read()` - One single function that reads object from the file cache

Some (minor) things I changed outside of `OsmHelper`:
1. Refactored `add_shape()` in `Route` class, to not query for each shape again and again (part of #2)
2. Removed `get_all_data()` from `Stop`. All data is queried at once and this function is obsolete (part of #2)
3. Moved `from_relation_member()` function from `Stop` class to `_build_stop()` in OsmHelper class  (part of #31)
4. Moved `get_interim_stop_name()` from `Stop` to `StopsCreatorFenix`, because this is custom Fenix code.
5. Temporarily removed the refresh on single route functionality from the `main` wrapper. This was not necessary but it makes review easier for this pull request. I'm happy to open an additional issue for this and re-implement this functionality once this pull request went in. And: refreshing all routes is working faster than one route before, due to the query optimizations... :smile:

In order to get rid of `OsmApi` library in favour of using exclusively `Overpy` we have to use a more recent version of overpy which has not been released officially, yet. But the [important commit](https://github.com/DinoTools/python-overpy/commit/ac7df97ef4302d7eda46a6bb34a887e9b275748c) went in more than one year (!) ago Please make sure that your get the new one. I had to do the following:

```
sudo pip uninstall overpy
sudo pip install -r requirements.txt
```

Make sure you re-cache data by running first:
`python osm2gtfs.py -c fenix.json.example --refresh-all
`
And then:
`python osm2gtfs.py -c fenix.json.example`

For me it runs through for Fenix (yeah! for the first time, since I'm working on this ), but validates that some stops are not used at all. Was this the case before? Maybe you can provide me (privately) with a last run of a resulting GTFS (`br-floripa.zip`), to be able to compare them. This can also be caused by the new queries and the custom blacklisted routes of Fenix code.

It is completely understandable if this bigger pull-request does not go in very quickly and I hope that still everybody is still motivated to test it, give feedback and make it a good progress. Personally I think this is one of the most important steps, as it makes it testable for me (finally). It improves huge parts of data retrieval, caching and readability of the whole process. I hope you also like it. Thanks a lot! Sorry again.

grote commented:
Wow, Felix! Thanks a lot! After browsing over the changes, this looks really good. I will probably only find time on the weekend to look into this in more detail.

xamanu commented:
Wonderful, I'm looking forward to hearing your feedback then! Thanks!

xamanu commented:
I did a couple of changes for better readability of the code:

* Moved caching functionality into a separate `Cache` class.
* Renamed `OsmHelper` to `OsmConnector`.

grote commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+
+            /* Return tags for elements */
+            );out body;""" % (self.tags, self.bbox)
+        return api.query(query_str)
```
This might be the reason why you are getting stops that are not part of any route. There's probably stops that are just not assigned to any route relation within the bounding box.

Doesn't need to happen within this MR, but it would be nice if we could check for those stops before the feed is validated and print useful error messages, so the user has a list of stops to look into.

xamanu replied:
The query is actually looking for relations (the same relations used for the routes based on the tags) and then downloads all objects with the role "platform" of these relations (no matter which tags).

We could think about altering this to download all nodes which are tagged as possible stops. And then try to merge them with the route relations. But what about stops that are not properly tagged, what about stops that are from other networks, not subject of this. I think the common ground is, that we download osm route relations and it's stops and then verify them.

xamanu replied:
This is correct. But there are also other roles, or object that should be in there without roles, etc. Personally I think we should encourage people to fix their OpenStreetMap data and not assume all possible failures they are doing.

BTW for poorly maintained routes nobody has schedule information and is using this script to generate GTFS. It's just easier to fix OSM data and better for everybody.

grote commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+
+            /*  Query for related bus stop platforms */
+            node(r:"platform")->.nodes;
+            way(r:"platform");
```
I am afraid not all stops in Floripa are using the platform tag. There's probably still many just using highway=>bus_stop. Those will just not be added with this query? If yes, how can we deal with this? Mass-add the proper tags to all stops that are still missing it?

xamanu replied:
First, there is a misunderstanding: This is not querying for platform tags, but the platform role in the relation.

grote replied:
Ahh, still need to learn this new syntax... Thanks for explaining.

grote commented on osmhelper/osm_routes.py:
```diff
@@ -56,6 +57,7 @@ def __repr__(self):
         rep += "http://www.consorciofenix.com.br/horarios?q=" + str(self.ref)
         return rep
 
+    # TODO: Move over to Fenix implementation
```
I can do those once we have your MR merged.

xamanu replied:
Cool!

grote commented on osmhelper/osm_routes.py:
```diff
@@ -110,49 +112,57 @@ def get_first_alt_stop(self):
+                sys.stderr.write(
+                    "Route has non-matching ways: " + str(self) + "\n")
+                sys.stderr.write(
+                    "  Problem at: http://osm.org/way/" + str(way) + "\n")
```
Does this work with https as well? If so, we should probably use that.

grote commented on osmhelper/osm_routes.py:
```diff
@@ -110,49 +112,57 @@ def get_first_alt_stop(self):
-        for nodes_sublist in [nodes[i:i + size] for i in range(0, len(nodes), size)]:
-            nodes_info = osm_api.NodesGet(nodes_sublist)
-            for node in nodes_sublist:
-                self.shape.append({'lat': nodes_info[node]["lat"], 'lon': nodes_info[node]["lon"]})
```
Awesome we can get rid of this! :)

xamanu replied:
yeah :smile:

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+            self.tags += str('["' + key + '" = "' + value + '"]')
+        else:
+            # fallback
+            self.tags = '["public_transport:version" = "2"]'
```
> print("No tags found for querying from OpenStreetMap.")
> print("Using tag 'public_transport:version=2")

grote replied:
sounds good!

xamanu replied:
Fixed in the updated push.

grote replied:
Thanks! I noticed that the fenix config file actually has tags defined, but they are not used. There must be a bug here as well which might also cause some routes from missing that don't have the public_transport:version tag. Let's use the tags from the config file for now and change it later to public_transport:version.

xamanu replied:
Thanks for reporting. There seemed to be an error with the commit before this PR. Should be fixed now with the latest commit update.

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+            rm = self._build_route_master(route_master, members)
+            # Make sure ref number is not already taken
+            if rm.ref in self.routes:
+                sys.stderr.write("Ref:" + rm.ref + "already in routes\n")
```
What about:

> sys.stderr.write("'Ref' of route_master already taken")
> sys.stderr.write("http://osm.org/relation/" + str(route_master.ref))
> sys.stderr.write("Skipped. Please fix in OpenStreetMap")

grote replied:
So the problem here is that there are two route master returned from the query with the same ref?

If so, the error message should say that I think. It should list both relations with link to OSM, so the user can quickly find both relations.

grote replied:
Also, please don't forget the line breaks when writing to STDERR

xamanu replied:
Fixed and added line breaks.

xamanu replied:
Added a [new comment ](https://github.com/grote/osm2gtfs/issues/30#issuecomment-258695414) about the test improvements to #30

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+            rv = self._build_route_variant(route_variant, result)
+            # Make sure ref number is not already taken
+            if rv.ref in self.routes:
+                sys.stderr.write("Ref:" + rv.ref + "already in routes\n")
```
> sys.stderr.write("Route (variant) with existing 'Ref'")
> sys.stderr.write("http://osm.org/relation/" + str(route_variant.ref))
> sys.stderr.write("Skipped. Please fix in OpenStreetMap")

xamanu replied:
Same to the comment above. All this proper error handling would be soo much easier within #30. So, let's not forget it, but also not make it part of the scope of this PR. Before the problem existed the same way, but not even any message was advised.

xamanu replied:
Added a [new comment ](https://github.com/grote/osm2gtfs/issues/30#issuecomment-258695414) about the test improvements to #30

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+                    otype = "way"
+
+                else:
+                    otype = "unknown"
```
We use it right below and it's important!

> stops.append(otype + "/" + str(stop_candidate.ref))

It's needed because osm ids are not unique over nodes and ways. So, the id's we use are like `node/329857` and `way/329857`

grote replied:
When you throw an exception, the code below isn't going to be executed.

xamanu replied:
Yes, I'm aware of this. I never expect here to be something else than way nor node. `unknown` would be indeed a serious issue.

grote replied:
My point is just, you don't need to assign otype anymore right before throwing a runtime exception, because the assignment isn't going to be used anyway ;)

xamanu replied:
Oh, you are very right :smile: That was not making any sense. Fixed now.

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+            (stop.lat, stop.lon) = Stop.get_center_of_nodes(stop.get_nodes())
+
+        s = Stop(stop.id, "node", stop.tags['name'], stop.lat, stop.lon)
+        print s
```
I'm happy to take it away. :smile:

xamanu replied:
Fixed.

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+        # Query relations of route variants, their masters and geometry
+        api = overpy.Overpass()
+        query_str = """(
+            /* Obtain route variants based on tags and bounding box */
```
:+1:

xamanu commented on creators/Fenix/StopsCreatorFenix.py:
```diff
@@ -1,19 +1,142 @@
+            # If there is no name, query one intelligently from OSM
+            if stop.name == Stop.NO_NAME:
+                self._get_interim_stop_name(stop)
+                print stop
```
Yes, actually this is the reason why I moved (recently) the cache into it's own class, so we can easily cache it here. I add a line to the `StopCreatorFenix`

```
if stops_changed:
     # Cache stops with newly created stop names
     Cache.write_data('stops', self.stops)
```

grote replied:
Thanks!

xamanu commented on creators/Fenix/StopsCreatorFenix.py:
```diff
@@ -1,19 +1,142 @@
+        # Add loose stop objects to route objects
+        self.add_stops_to_routes(schedule, data)
+
+    def add_stops_to_routes(self, schedule, data):
```
Yes, no need to have it here.

xamanu replied:
Fixed.

xamanu commented on creators/Fenix/StopsCreatorFenix.py:
```diff
@@ -1,19 +1,142 @@
+            for stop in route.stops:
+                if stop in stops:
+                    # Replace stop id with Stop objects
+                    route.stops[i] = stops[stop]
```
Yes, I'm with you. There is a basic new concept here. Let me explain:

Before: A route had it's stops as objects assigned. One identical stop, which has been used by different routes was instantiated several times and was handled as different ones. I didn't like the idea. 

Now: With the query optimizations this is not happening anymore. Routes and Stops are handled separately and efficiently. The route objects usually have a dictionary of (only) stop ids to point to the stop objects.

But Fenix current logic expects them to be objects. I would suggest to change Fenix' logic in the future. This replacement from ids to objects is just a temporary way to simulate the conversion which actually should be improved (in my humble opinion) in Fenix' logic.

grote replied:
What you call an idea was completely unintentional ;)

I'm fine leaving it like this for now, but please add an TODO that explains what needs to be done.

xamanu replied:
ok. Added in latest update of the commit.

xamanu commented on creators/Fenix/StopsCreatorFenix.py:
```diff
@@ -1,19 +1,142 @@
+        else:
+            raise RuntimeError("Unknown Route: " + str(route))
+
+    def _get_interim_stop_name(self, stop):
```
Yes, those are the ones that make it (still) slow. But an optimization of this is clearly out of scope of this PR. This should be handled in a separate issue and though through very well. Then this could be also interesting not only for Fenix but also other networks.

grote replied:
Yes, the optimization is out of scope, but renaming the method isn't.

xamanu replied:
Yes, you are right. Fixed in latest update of commit.

xamanu replied:
Here is the ticket for optimization: #39 (Find best name for unnamed stop)

xamanu commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,325 @@
+                                ] = self._build_route_variant(rv, result)
+                    else:
+                        print "Member relation is not a valid route variant:"
+                        print "http://osm.org/relation/" + str(member.ref)
```
Fixed in the updated push.

grote commented on creators/Fenix/StopsCreatorFenix.py:
```diff
@@ -1,19 +1,150 @@
+
+        if stops_changed:
+            # Cache stops with newly created stop names
+            Cache.write_data('stops', self.stops)
```
I actually meant to cache the stops after every call to `find_best_name_for_unnamed_stop()`. This might seem excessive, but disk access is still way fast than network access. When running your branch, the script terminates quite often when finding the names and with each restart it needs to start from scratch. Would be nice if it could continue where it left off.

grote replied:
Also, what does `self.stops` refer to here?

xamanu replied:
Yes, good idea to cache every time a name has been found (successfully or not), to be able to continue. This also would allow to get rid of the `stops_changed` flag. I love it. Changed in latest update to the commit.

grote commented:
I am still running a test for a version from some commits earlier. Has the missing trips issue already been addressed? If not, do you need help finding it?

xamanu commented:
Thanks for all the good feedback! No, I wasn't able to find and address properly the missing trips issue. Your help on this would be very much appreciated! Hopefully you are more efficient on this, with the Fenix code.

xamanu commented on osmhelper/osm_stops.py:
```diff
@@ -1,11 +1,7 @@
-from transitfeed import util
-
 
 STOP_REGEX = re.compile('(TICAN|TISAN|TICEN|TITRI|TILAG|TIRIO|TISAC).*')
```
Hmm, but `normalize_name` is only getting called from `match_first_stops` which only get's called from `TripsCreatorFenix` when everything is already build as it should be. If we start working on this, it would already be related very much to #30... And I wanted to make this PR small :laughing:

grote replied:
Sure, let's keep this small and don't change it any more than really necessary. I am just trying to find the reason, why the trips aren't created. And so far it looks like the stops can't be matched due to the normalization failing. Maybe that is relating to the new way of storing the stop objects?

xamanu replied:
This seems to be solved by now.

grote commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,336 @@
+
+                stops.append(otype + "/" + str(stop_candidate.ref))
+
+        rv = Route(route_variant.id, None, None, stops, None, ref)
```
I think this is the bug. You are just passing None in to and from, but I need those to identify the route variants.

grote replied:
The name though not as critical is also `None` and can probably already be set here.

xamanu replied:
Done.

grote commented on osmhelper/osm_helper.py:
```diff
@@ -1,328 +0,0 @@
-        if ref in route_masters:
-            # route has a master, so update it in the route_master and add it to the routes
-            stops = route_masters[ref].routes[rid].stops
-            route_masters[ref].routes[rid] = Route(rid, fr, to, stops, route_masters[ref], ref, name)
```
That looks like the critical code you dropped and that is most likely causing the bug.

xamanu replied:
I made some changes (recent commit update), now the gtfs is way bigger (1.7MB), but probably not as big as the original outcome (2.4MB)

xamanu replied:
I get 5501 trips... Have you done `--refresh-routes` or `--refresh-all`?

xamanu replied:
Could this be because oft he missing `RouteMaster` object in the same `Route`(variant) initiation? It's just anoying, because I'm building the `RouteMaster` object after creating the `Route`(variants)...

grote replied:
Check out this route: http://www.openstreetmap.org/relation/2079729
I'm getting this for this route:
```
Route (variant) with existing 'Ref'
http://osm.org/relation/2079727
Skipped. Please fix in OpenStreetMap
```
What is the problem with that route?

xamanu replied:
Got it! I was actually replacing `Route`(variants) inside a `RouteMaster` when having the same `ref`. This was not correct and has been fixed in the latest commit update. Thank you for spotting me that! Now I'm at 9092 trips.... That are too much? Was the old code not generating all trips? Or is there still a problem?

grote replied:
Wohoo that sounds good! :) Good catch! :+1: 

Let me run that myself and check if there's any outstanding issues.

grote replied:
Did you push that fix? It is still skipping routes like this one http://www.openstreetmap.org/relation/1864112

xamanu replied:
Eh, yes. See line 117 of osm_connector.py
>  members[rv.id] = self._build_route_variant(rv, result)

xamanu replied:
I also have `route_short_name` (`ref`) 843 in routes.txt (`route_id`=69) and 116 trips in trips.txt

grote replied:
Is your HEAD a2be0b3f88cc5a5e861f536199d8de06975792b6 ?

xamanu replied:
Yes. Want to check on [Mumble](http://wiki.openstreetmap.org/wiki/Mumble)?

grote commented on osmhelper/osm_connector.py:
```diff
@@ -0,0 +1,351 @@
+            if rv.ref in self.routes:
+                sys.stderr.write("Route (variant) with existing 'Ref'\n")
+                sys.stderr.write(
+                    "http://osm.org/relation/" + str(route_variant.ref) + "\n")
```
```
Traceback (most recent call last):
  File "./osm2gtfs.py", line 111, in <module>
    main()
  File "./osm2gtfs.py", line 58, in main
    data.get_routes(refresh=True)
  File "osm2gtfs/osmhelper/osm_connector.py", line 153, in get_routes
    "http://osm.org/relation/" + str(route_variant.ref) + "\n")
AttributeError: 'Relation' object has no attribute 'ref'
```

grote replied:
`s/route_variant/rv/`?

xamanu replied:
Not sure, but I think it's `s/ref/id/` ? (route_variant.id)

grote replied:
crazy indeed. I think you are right and it should be route_variant.id

xamanu replied:
Fixed with the latest commit update.

grote commented on osmhelper/osm_stops.py:
```diff
@@ -33,123 +29,7 @@ def __repr__(self):
+    # TODO: Move over to Fenix implementation
     @staticmethod
     def normalize_name(old_name):
         name = STOP_REGEX.sub(r'\1', old_name)
```
Could you please add this here as a temporary fix:
```python
name = name.replace('Terminal Centro', 'TICEN')
```
Somebody has renamed this station and broke the normalization https://www.openstreetmap.org/way/43719878#map=18/-27.59847/-48.55306

This gets us some more trips back.

xamanu replied:
Added with the latest update on the commit.

xamanu commented:
Ah, I just saw I wrote Thorsten instead of Torsten. Sorry for that!!
